driver-dse/src/main/java/com/datastax/driver/auth/DseAuthProvider.java

```diff
@@ -18,67 +18,176 @@
+ * The mechanism is sent to the server in an exchange prior to the SASL exchanges. If
+ * the server is configured to support the mechanism then it replies with <code>mechanism-START</code>.
  * <p/>
- * Keytab and ticket cache settings are specified using a standard JAAS
- * configuration file. The location of the file can be set using the
- * <code>java.security.auth.login.config</code> system property or by adding a
- * <code>login.config.url.n</code> entry in the <code>java.security</code> properties
- * file.
- * <p/>
- * See the following documents for further details on the
- * <a href="https://docs.oracle.com/javase/6/docs/technotes/guides/security/jgss/tutorials/LoginConfigFile.html">JAAS Login Configuration File</a> and the
- * <a href="http://docs.oracle.com/javase/6/docs/technotes/guides/security/jaas/tutorials/GeneralAcnOnly.html">JAAS Authentication Tutorial</a>
- * for more on JAAS in general.
- * <p/>
- * <h1>Authentication using ticket cache</h1>
- * Run <code>kinit</code> to obtain a ticket and populate the cache before
- * connecting. JAAS config:
- * <p/>
- * <pre>
- * DseClient {
- *   com.sun.security.auth.module.Krb5LoginModule required
- *     useTicketCache=true
- *     renewTGT=true;
- * };
- * </pre>
- * <p/>
- * <p/>
- * <h1>Authentication using a keytab file</h1>
- * To enable authentication using a keytab file, specify its location on disk.
- * If your keytab contains more than one principal key, you should also specify
- * which one to select.
- * <p/>
- * <pre>
- * DseClient {
- *     com.sun.security.auth.module.Krb5LoginModule required
- *       useKeyTab=true
- *       keyTab="/path/to/file.keytab"
- *       principal="user@MYDOMAIN.COM";
- * };
- * </pre>
- * <p/>
- * To connect to clusters using internal authentication, use the standard method for
- * setting credentials. eg:
- * <p/>
- * <pre>
- * Cluster cluster = Cluster.builder()
- *                          .addContactPoint(hostname)
- *                          .withCredentials("username", "password")
- *                          .build();
- * </pre>
+ * This is an abstract class. To make use of this auth provider one of the following should be used:
+ * <ul>
+ * <li>{@link DsePlainTextAuthProvider} for PLAIN SASL authentication</li>
+ * <li>{@link DseGSSAPIAuthProvider} for GSSAPI SASL authentication</li>
+ * </ul>
  */
-public class DseAuthProvider implements AuthProvider {
-    @Override
+public abstract class DseAuthProvider implements AuthProvider {
+    private static final String DSE_AUTHENTICATOR = "com.datastax.bdp.cassandra.auth.DseAuthenticator";
+    private final String username;
+    private final String password;
+
+    public DseAuthProvider(String username, String password) {
+        this.username = username;
+        this.password = password;
+    }
+
     public Authenticator newAuthenticator(InetSocketAddress host, String authenticator) throws AuthenticationException {
```

Comment:
Sorry for the spam :) but since we are modifying the API there should also be an entry in `upgrade_guide/README.md` to explain how to migrate existing code, can you add that too? I guess it boils down to replacing `DseAuthProvider` with `DseGSSAPIAuthProvider`. Thanks!